### ISSUE TITLE
No tests run when nested classes in generic fixture.

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#2789]
_From @JohanLarsson on March 30, 2018 14:16_

```cs
using NUnit.Framework;

[TestFixture(typeof(int))]
[TestFixture(typeof(double))]
public partial class Foo<T>
{
    public class Bar
    {
        [Test]
        public void BarTest()
        {
        }
    }
}

public partial class Foo<T>
{
    public class Baz
    {
        [Test]
        public void BazTest()
        {
        }
    }
}
```

The VS runner does not run any test from what I can see.

Visual Studio Enterprise 2017 v15.6.4
NUnit 3 Test Adapter v3.10.0.0
NUnit v3.9.0
net461

_Copied from original issue: nunit/nunit3-vs-adapter#498_

**Comment by rprouse** [comment]
Have you attempted to run this using the console? If so, does it work? This will help us determine if the bug is in Visual Studio only, or actually in the framework.

**Comment by rprouse** [comment]
Also, do the tests show up in the Test Explorer window?

**Comment by rprouse** [comment]
_From @JohanLarsson on March 30, 2018 18:34_

I did not run it with console.
Does the code not reproduce the issue?

**Comment by rprouse** [comment]
The code may reproduce the issue, but I deal with between 20 or 30 issues a day across the various NUnit projects, all in my spare time, so reproducing all issues takes time and we can use all the help we can get 😄

**Comment by rprouse** [comment]
_From @JohanLarsson on March 30, 2018 19:38_

Visual Studio runner:
![animation](https://user-images.githubusercontent.com/1640096/38150551-bfa91aa4-345f-11e8-8a5b-49111b907511.gif)

Console runner:
```
Microsoft Windows [Version 10.0.16299.309]
(c) 2017 Microsoft Corporation. All rights reserved.

C:\Users\Johan\Downloads\NUnit.Console-3.8.0>nunit3-console.exe C:\Git\Kasta\WpfApp1\WpfApp1.Tests\bin\Debug\WpfApp1.Tests.dll
NUnit Console Runner 3.8.0
Copyright (c) 2018 Charlie Poole, Rob Prouse

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.16299.0
  CLR Version: 4.0.30319.42000

Test Files
    C:\Git\Kasta\WpfApp1\WpfApp1.Tests\bin\Debug\WpfApp1.Tests.dll


Run Settings
    DisposeRunners: True
    WorkDirectory: C:\Users\Johan\Downloads\NUnit.Console-3.8.0
    ImageRuntimeVersion: 4.0.30319
    ImageTargetFrameworkName: .NETFramework,Version=v4.6.1
    ImageRequiresX86: False
    ImageRequiresDefaultAppDomainAssemblyResolver: False
    NumberOfTestWorkers: 4

Test Run Summary
  Overall result: Passed
  Test Count: 0, Passed: 0, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2018-03-30 19:36:13Z
    End time: 2018-03-30 19:36:15Z
    Duration: 1.523 seconds

Results (nunit3) saved as TestResult.xml
```

I never use the console runner so not sure I did it right.

**Comment by rprouse** [comment]
Your run looks good. Can you share the test project?

**Comment by rprouse** [comment]
_From @JohanLarsson on March 30, 2018 19:49_

https://github.com/JohanLarsson/NUnitRepro

Place I found it was after refactoring [this](https://github.com/DotNetAnalyzers/IDisposableAnalyzers/blob/master/IDisposableAnalyzers.Test/IDISP003DisposeBeforeReassigningTests/HappyPath.cs#L9) to be a parametric fixture.

**Comment by rprouse** [comment]
_From @OsirisTerje on March 30, 2018 20:0_

I can reproduce the issue.  I wonder if this might be a variation around 1)  Issue #218 (No tests found to run when TestFixture is a nested class) which was fixed late 2016 for version 3.7, but did not check generics.

I also see that the new source based discovery (SBD) finds the tests, but won't run them as reported.  If I turn off SBD, then the adapter doesn't find the tests either. 

Running it from R# unit tests, they are found and executed, so this must be related to the adapter itself. 

Testing a bit more:
1. A partial class with no generics:  Works
2. A non-partial class with generics:  Does **not** work

**Comment by rprouse** [comment]
_From @JohanLarsson on March 30, 2018 20:1_

R# is shaky, runs them sometimes.

**Comment by rprouse** [comment]
_From @OsirisTerje on March 30, 2018 20:11_

@JohanLarsson Interesting, there might be something further down here.  When I enable discoverydump, I see that the GenericFixture and the two internal classes are found, but not any test methods.
`  <test-suite type='GenericFixture' id='0-1002' name='Foo&lt;T&gt;' fullname='Partial.Foo&lt;T&gt;' runstate='Runnable' testcasecount='0'>
            <test-suite type='TestFixture' id='0-1001' name='Foo&lt;Double&gt;' fullname='Partial.Foo&lt;Double&gt;' classname='Partial.Foo`1' runstate='Runnable' testcasecount='0' />
            <test-suite type='TestFixture' id='0-1000' name='Foo&lt;Int32&gt;' fullname='Partial.Foo&lt;Int32&gt;' classname='Partial.Foo`1' runstate='Runnable' testcasecount='0' />
         </test-suite>`

@rprouse This means we don't get anything up from NUnit/Engine itself.  Further, the non-partial generic class is not coming up at all.

**Comment by rprouse** [comment]
_From @OsirisTerje on March 30, 2018 20:19_

If I add the TestFixture attribute to the inner class, it **is** discovered, but not reported, because the fixture is skipped.
` 
     [TestFixture(typeof(int))]
     [TestFixture(typeof(double))]
     public partial class Foo<T>

     {

        [TestFixture]

        public class Bar

        {

            [Test]

            public void BarTest()/

            {
            }

        }
    }
`

Gives the following output:

` 
<property name='_SKIPREASON' value='Fixture type contains generic parameters. You must either provide Type arguments or specify constructor arguments that allow NUnit to deduce the Type arguments.' />
                 `

![image](https://user-images.githubusercontent.com/203432/38152579-d216167a-3468-11e8-85ff-e86eeb34ba7c.png)

**Comment by rprouse** [comment]
_From @jnm2 on March 31, 2018 20:12_

It looks like the framework is not using the `[TestFixture(typeof(x))]` attributes on the containing class to satisfy the generic parameter in the nested class. (Nested types inherit all type parameters from containing types.)

**Comment by rprouse** [comment]
The fact that the run from NUnit console resulted in no tests run leads me to believe that it is an NUnit issue. This issue should be in the nunit repo, moving.

**Comment by jnm2** [comment]
This has nothing to do with partial classes, just classes declared inside generic classes.

The problem in general is shockingly complex to solve and will result in breaking changes to the IFixtureBuilder interface. For example, generic argument inference from constructor argument values needs to constrain generic arguments declared in the containing class. That then limits which generic instantiations the containing class multiplies the nested class with.

**Comment by rprouse** [comment]
Thanks for looking at this @jnm2. Based on the difficulty of fixing this and the fact that it is an edge case, I think the priority to fix should be fairly low. @JohanLarsson can you make a use case for this use case being higher priority? What does it allow you to do that restructuring your tests can't accomplish?

**Comment by CharliePoole** [comment]
Since it doesn't work at all (correct?) and is low-use anyway, perhaps an error would be better than an analyzer. We could simply treat such fixtures as non-runnable.

**Comment by JohanLarsson** [comment]
Yes, that is a very reasonable.

**Comment by rprouse** [comment]
I am going to increase the priority with the assumption that we will **treat it as an error** and report the tests as unrunnable.

**Comment by JohanLarsson** [comment]
Maybe a separate issue for treat as error with high prio?

**Comment by jnm2** [comment]
The problem itself is one I would very much enjoy solving, but it's a matter of prioritizing keyboard time.

**Comment by AGBrown** [comment]
This bit me today. I have a generic TestFixture with a nested (and therefore generic) fixture. It isn't showing in VS 2013 test discovery (NUnit 3.7.1, NUnit3TestAdapter 3.10.0), nor in R# test lists.

```
[TestFixture(typeof(MyTypeA))]
[TestFixture(typeof(MyTypeB))]
public class MyTypeFixture<T>
  where T : IMyServiceType
{
  [TestFixture]
  public class MyMethodOneFixture
  {
    // test methods here for IMyServiceType.MyMethodOne
  }
  [TestFixture]
  public class MyMethodTwoFixture
  {
    // test methods here for IMyServiceType.MyMethodTwo
  }
}
```

The use case for nexted fixtures is to help segment code. E.g. if a method in a SUT has a lot of tests, instead of prefixing all the test names with `MyMethod_`, they just get dumped in a nested fixture `MyMethodFixture`, resulting in shorter test names which are easier to scan.

The use case for a generic test fixture is when multiple classes expose methods that behave exactly the same for each class, and they share a supertype or interface which becomes the SUT. I have to admit that we have very few use cases where this applies, but it would definitely be useful to combine the two test patterns where we need it.

**Comment by CharliePoole** [comment]
Can you replicate this in the latest framework release? Can you replicate it running with either the console runner or through nunitlite?

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw 
This issue is old (2018/19), but marked High Pri, although the analysis done back then says it is **very hard to fix**, but we could raise an error for this case, OR, we could detect this with the analyzer and warn that it will not be discovered and thus not run.


@jnm2 said:

> It looks like the framework is not using the [TestFixture(typeof(x))] attributes on the containing class to satisfy the generic parameter in the nested class. (Nested types inherit all type parameters from containing types.)

@rprouse said

> I am going to increase the priority with the assumption that we will treat it as an error and report the tests as unrunnable.

I made a repro for this with the latest 5.0 alpha  version here: https://github.com/nunit/nunit.issues/tree/main/Issue2789  and verified, with `dotnet test  -s ".runsettings` that there are no changes up to version 5 that have touched this.

When enabling the dump, one can see the SKIP reason on the fixture (!) 

```xml
<property name='_SKIPREASON' value='Fixture type contains generic parameters. You must either provide Type arguments or specify constructor arguments that allow NUnit to deduce the Type arguments.' />
```
But this message doesn't appear in the output, so that is another miss.

Execution dump file enclosed.

[E_Issue2789.dll.zip](https://github.com/user-attachments/files/30201621/E_Issue2789.dll.zip)

The framework finds the fixtures, but not the test methods within.

**Comment by manfred-brands** [comment]
@OsirisTerje I will have a look to see how inner type fixtures are created

**Comment by stevenaw** [comment]
This may be easier now that `TestFixture` supports the `TypeArgs` property. Those are intended to be analogous to explicitly specifying the generic types at compile time and should remove the need for type inference or inspection of the nesting type.

**Comment by manfred-brands** [comment]
> But this message doesn't appear in the output, so that is another miss.

@OsirisTerje I have seen at work sometimes the 'no test' error, but never the reason. Showing this normally would be appreciated.

**Comment by manfred-brands** [comment]
> The framework finds the fixtures, but not the test methods within.

The framework doesn't look for `IFixtureBuilder`  attributes on the outer type.

Nothing in the [TestFixture documentation](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testfixture.html?q=TestFixture) mentions anything about nested classes and how these are supposed to work.

There is an inconsistency between the documentation and the code, before I correct one of them how do we want this to work:.
The [inheritance section](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testfixture.html#inheritance)
seems to indicate that attributes from base classes are combined.

This is not the case, if a class has _any_ attributes, the search stops and doesn't look in base classes.
This might be intentional.

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw 

In such conditions, confusion/contradiction and so on, I try to think of the `least surprise` principle.  Aka to avoid the `Design by surprise`.  

So:

If I put an attribute on the base class, would I expect that to work on the concrete derived class?  Or would I be **surprised** if it didn't? Or would I be surprised if it did?
I think I would be surprised if it didn't. 

Now, a nested class.

Would I be surprised if the inner classes wasn't regarded as test fixtures too (given the examples in the beginning and the code in the issue repo) ?  
Yes, I absolutely would.

In the code the test would have run, if it was **not** encapsulated by the parametric test fixture class.  

In the repro, the VS RTD finds the tests.  
It it also reports a misleading warning:  

<img width="800" height="27" alt="Image" src="https://github.com/user-attachments/assets/e4f4705a-98da-4959-bce5-4928335d178f" />

So, as a user,  I am getting surprised, and confused - because I don't know what to do here.  

Now, from a C# perspective, everything (not really true) in the outer class is visible to the inner class.  But - attributes declared on the outer class doesn't apply to the inner class.

So, from a C# design perspective,  the attributes should not go further by themselves, but their parameters would do so, if declared, and they are.  

But what happens in our case is not that the attributes need to pass down,  but that the tests that are declared are being ignored.  The use the TestFixture attribute on a fixture is optional, and even **with** the attribute on the inner class , they are not discovered nor executed by NUnit. 

So yes, I think we should fix this.

**Comment by CharliePoole** [comment]
I agree. All the stuff about the attribute passing down is really not relevant. All classes, nested are not, are examined in the same way, so if a class has tests that would be found if it were _not_ nested, then they should be found if it's nested.

However, _if_ the TestFixture attribute is being handed down to the nested class (which it should not be) then maybe the problem is that the nested class is not generic and then the framework doesn't know what to do with it.

BTW, I think I'm the one who used the word "inherited" in the docs and I regret it. It's confusing to say that the attribute is "inherited" in a section about actual class inheritance because the only thing that's "inherited" is the fact that the class is a TestFixture. So something like "Any inherited classes are considered TestFixtures." But they don't become parameterized fixtures or generic fixtures unless you actually write them that way and then you need to provide the proper form of the attribute.

**Comment by manfred-brands** [comment]
A nested class inside a generic class cannot be instantiated without knowing the generic parameters of the outer class and hence cannot be treated as being independent of the `TestFixtureAttribute` on that outer class.

My PR does do that use the generic parameter from the outer class.

**Comment by OsirisTerje** [comment]
Sorry @manfred-brands - I missed your PR (and now see I also commented there.......sighs)

Long discussion in the PR - haven't read everything there, but if you're happy about it, - can we just merge it in ?

**Comment by stevenaw** [comment]
@OsirisTerje @manfred-brands 

I'm also interested in seeing and helping this move forward.
I'd left a few comments in the PR hoping we could find a solution which didn't require consumers to implement or even know about the newly proposed public interface methods. Earlier in that discussion I'd also suggested we merge an earlier draft of the PR which solved only the simple cases and deferred fixing the more complex cases to give a bit more time to find ways to solve them as unobtrusively as possible

**Comment by OsirisTerje** [comment]
@stevenaw Can you summarize your comments in the PR,  I am drowning in all the discussion there, so I can't actually find which comments are something we (or Manfred) should act on.

I agree it is fine deferring complex cases for later.

**Comment by stevenaw** [comment]
@OsirisTerje I think as of this commit could be one place where we've made the simple case outlined in the above repro both work
https://github.com/nunit/nunit/pull/5370/changes/e880f2823e9a5f9f16ae7a825ba1c1dab2c776cf

We could potentially merge from that point, perhaps with the exception that we'd also want to first add test coverage for the other "forms" as described here since it sounds like they "just work" https://github.com/nunit/nunit/pull/5370#discussion_r3630331216


This more complex scenario where _each_ the outer and inner class separately define generic types is where I believe it gets complicated
https://github.com/nunit/nunit/pull/5370#discussion_r3642591355
The complication was solved by extending a new `IFixtureBuilder3` from the existing `IFixtureBuilder2` in order to add a few new public methods for nested generic fixtures. This is the piece I think is getting caught up in some discussion, though in my opinion it needn't hold up the simple case above from merging if @manfred-brands is in agreement https://github.com/nunit/nunit/pull/5370#discussion_r3744778703

**Comment by manfred-brands** [comment]
@stevenaw The commit you are referring to was flawed in the sense that if the inner class had a `[TestFixture]` attribute, it wouldn't work because the code would find an `IFixtureBuilder` on the nested class and not look for the parent one.
It is hard to explain to users that an optional attribute is the reason it doesn't work.

I have fixed that and pushed it to the PR.
This only fixes the non-generic-non-parameterized test inside a generic class.

Note that I think that the existing filtering in  `GetFixtureBuilderAttributes` is flawed as it reduces multiple user defined attributes to one as it assumes those cannot have arguments. Although I cannot find many user implementations on github, they could be.

Note that having an attribute means the fixture is built using `TestFixtureAttribute.BuildFrom` whilst not having one means going directly  through `NUnitTestFixtureBuilder.BuildFrom`, similar code slightly different paths.

I will create a separate PR on top of the one above to deal with nested generic classes.
